docs/spec/concepts.rst

```diff
@@ -3,97 +3,216 @@
+requirement, but it's a useful guideline.
+
+The subtype relation
+--------------------
+
+A static type ``B`` is a **subtype** of another static type ``A`` if and only
+if the set of values represented by ``B`` is a subset of the set of values
+represented by ``A``. Because the subset relation on sets is transitive and
+reflexive, the subtype relation is also transitive (if ``C`` is a subtype of
+``B`` and ``B`` is a subtype of ``A``, then ``C`` is a subtype of ``A``) and
+reflexive (``A`` is always a subtype of ``A``).
+
+The **supertype** relation is the inverse of subtype: ``A`` is a supertype of
+``B`` if and only if ``B`` is a subtype of ``A``; or equivalently, if and only
+if the set of values represented by ``A`` is a superset of the values
+represented by ``B``. The supertype relation is also transitive and reflexive.
+
+We also define an **equivalence** relation on static types: the types ``A`` and
+``B`` are equivalent (or "the same type") if and only if ``A`` is a subtype
+of ``B`` and ``B`` is a subtype of ``A``. This means that the set of values
+represented by ``A`` is both a superset and a subset of the values represented
+by ``B``, meaning ``A`` and ``B`` must represent the same set of values.
+
+We may describe a type ``B`` as "narrower" than a type ``A`` (or as a "strict
+subtype" of ``A``) if ``B`` is a subtype of ``A`` and ``B`` is not equivalent
+to ``A``.
+
+The consistency relation
+------------------------
+
+Since :ref:`Any` represents an unknown static type, it does not represent any
+known single set of values, and thus it is not in the domain of the subtype,
+supertype, or equivalence relations on static types described above. (We do
+find it convenient to say, however, that ``Any`` is both subtype and supertype
+of -- that is, equivalent to -- only itself. This can allow us to simplify
+redundant multiple occurrences of ``Any`` out of more complex types.)
+
+To relate gradual types more generally, we define a **materialization**
+relation. The intuition for materialization is that it transforms a "more
+dynamic" type to a "more static" type. Given a gradual type ``A``, if we
+replace zero or more occurrences of ``Any`` in ``A`` with some gradual type
+(which can be different for each occurrence of ``Any``), the resulting gradual
+type ``B`` is a materialization of ``A``.
+
+For instance, ``tuple[int, str]`` (a static type) and ``tuple[Any, str]`` (a
+gradual type) are both materializations of ``tuple[Any, Any]``. ``tuple[int,
+str]`` is also a materialization of ``tuple[Any, str]``.
+
+If ``B`` is a materialization of ``A``, we can say that ``B`` is a "more
+static" type than ``A``, and ``A`` is a "more dynamic" type than ``B``.
+
+The materialization relation is both transitive and reflexive, so it defines a
+preorder on gradual types.
+
+We also define a **consistency** relation on gradual types.
+
+A static type ``A`` is consistent with another static type ``B`` if and only if
```

Comment:
Yeah, I think that's a loose/colloquial usage. It's a reasonable usage given the informal meaning of the term, but the meaning presented here is the one used in the literature, and I don't think there's a sufficiently strong reason to use something different. I don't think there will be very many places in the spec where we need the consistency relation at all. TypeVar constraints will be one.

Reply:
In some literature, such as Abstracting Gradual Types, the equivalence relation is lifted to the consistency relation. Are we saying that consistency can be thought of as consistent subtyping, since in practice, consistency doesn't actually come up that much since we are usually considering systems with subtyping? Right now the actual definitions look similar to what's in the literature.

Reply:
No.  According to AGT, lifting static typing **equivalence** gives gradual typing **consistency**; and lifting static typing **subtyping** gives gradual typing **consistent subtyping**.

The idea of AGT is that you design a fully statically typed language, which is "easy".  And then you get gradually typed equivalents of the static typing constructs "for free" by using their approach.

Reply:
PEP-483 loosely defines an "is-consistent-with" relation but it means consistent subtyping.  That has understandably led people to use the term "consistent" when they are talking about consistent subtyping, which could be confusing.

We almost don't need consistency at all, so it hasn't caused us any real problems to use the wrong term for consistent subtyping (other than constantly annoying me :) ).

Reply:
Makes sense! I also would immediately think of the standard consistency relation (instead of consistent subtyping) when seeing the consistency term discussed above. So the description above looks right to me. But I could also be one of a few people!